### PR TITLE
feat(engine): apply the CR 605.1a library-movement criterion

### PR DESCRIPTION
**Pull request by matthewevans**
## What

CR 605.1a was amended to add a fourth criterion to the mana-ability test:

> An activated ability is a mana ability if it meets all of the following criteria: it doesn't require a target (see rule 115.6), it could add mana to a player's mana pool when it resolves, it's not a loyalty ability (see rule 606, "Loyalty Abilities"), **and its cost and effect don't move any card to or from a library**. **Do not take into account replacement effects that may apply, other than self-replacement effects, when evaluating these criteria.**

Two deltas versus the prior printing: the library-movement criterion, and the trailing replacement-effect clause. (The amended text is already present in `docs/MagicCompRules.txt`.)

## How

- **`Effect::moves_card_to_or_from_library()` / `AbilityCost::moves_card_to_or_from_library()`** — exhaustive, wildcard-free per-node predicates over all **229** `Effect` and **33** `AbilityCost` variants. Every reasoned verdict carries its CR annotation inline. No wildcard arm, so a future variant cannot silently inherit `false`.
- **`ResolutionScope { OwnResolutionOnly, IncludeRegisteredLater }`** in `types/ability_visit.rs`. `OwnResolutionOnly` stops at effects that merely *register* a separate ability to apply later — delayed triggers (CR 603.7a), reflexive triggers (CR 603.12), registered replacements (CR 614.1), and `Effect::Mana`'s `grants` (CR 603.3). All 8 existing public entry points keep `IncludeRegisteredLater`, so no current caller changes behavior.
- **`is_mana_ability` is now the single authority for all four criteria.** The three-criteria core is extracted as `produces_mana_on_activation` so that `is_renewable_mana_ability` — which asks a different question, "is this permanent part of a standing manabase?" — does not narrow. Composing the development predicate on the rules predicate would drop Millikin, Deranged Assistant and Codie out of `phase-ai`'s mulligan `keep_tier` for a reason unrelated to manabase development.

The **replacement-effect clause is satisfied by construction**: `is_mana_ability` is a pure function of the printed AST, takes no `&GameState`, and therefore cannot consult replacement effects at all. Self-replacement riders printed on the ability (CR 614.15) — e.g. Memory Lapse's "put it on top of its owner's library instead", which is CR 608.2c's own worked example — are counted, which is what the closing carve-out requires.

Near-misses that deliberately stay classified, each for a different CR reason: Chromatic Star (separate `ChangesZone` trigger), Barbed Sextant (CR 603.7a), Shaun & Rebecca (CR 603.12), Gilanra (CR 603.3), The Secret Lair (CR 701.22a — scry reorders *within* a library).

## Verification

- Tilt `clippy` clean; `test-engine` **23,355 / 23,356 passing**. The single failure (`battlefield_entry_authority_census`) is another agent's concurrent in-flight work — that test is a source-scanning census over `TokenCreated` constructions and no path in this PR contains one.
- Pre-commit gates pass: parser-combinator **Gate A** + **Gate G**, PreLowered **Gate P**.
- Four new integration tests, deliberately asserting **no AST-internal flag** — they pin only what is observable by driving the real pipeline. Every negative is paired with a positive reach-guard *in the same test* so a fixture that never reached the seam cannot pass vacuously.
- The snapshot corpus acts as an independent cross-check: **16** committed snapshots pin `is_mana_ability`, the suite exercised all 16, and exactly one moved — Manamorphose, the one the census predicted.
- All CR citations were grep-verified against `docs/MagicCompRules.txt` before being written.

## Reviews

Plan reviewed to convergence (5 rounds, `review-r8`: APPROVED). Implementation independently reviewed: **12 findings, zero behavioral defects** — every arm verdict against the CR was correct. All 12 fixed in the second commit, notably:

- `Effect::Ripple`'s comment quoted only the reveal and bottom-of-library legs of CR 702.60a, both of which this file classifies as `false` — so it argued against its own `true`. The warrant is the **casting** leg (CR 702.60a + CR 601.2a, library → stack), now cited explicitly.
- CR 603.3 was cited as *primary* authority for the replacement-registration boundary at five sites. CR 603.3 is exclusively about a *triggered* ability going on the stack; replacement effects never trigger. Swapped to CR 614.1 / CR 614.15.

## Known divergences (documented, none pinned as correct by any test)

- **Codie, Vociferous Codex** is reclassified when CR 605.1a says it should not be. Accepted, documented divergence traced to a **parser mis-attachment** (`CastFromZone` / `PutAtLibraryPosition` parsed as chain siblings of the `CreateDelayedTrigger` rather than as its payload). Codie is *already* rules-incorrect today for the same root cause — its free-cast continuation resolves immediately instead of when the delayed trigger fires. A follow-up PR fixes the attachment, after which `OwnResolutionOnly` prunes them and Codie classifies correctly **with no further change here**.
- **Pre-existing CR 605.5b violation** (`is_mana_ability` never checks `AbilityKind`, so spell-kind abilities serialize `is_mana_ability: true`). This change incidentally flips exactly one of them, Manamorphose — right outcome, wrong criterion. 31 remain misflagged over top-level `.abilities[]` nodes. Needs its own census.
- **Chartered follow-up — cost walk does not descend inline branch carriers.** `visit_effect_scoped` descends nested `AbilityDefinition`s under `Vote`, `SeparateIntoPiles`, `RevealFromHand.on_decline`, `FlipCoin`/`FlipCoins`, `FlipCoinUntilLose`, `RollDie.results` and `ChooseOneOf.branches`; the cost walk does not. A library cost on such a branch would go unclassified. **Zero reachable instances today** (`card-data.json` carries no costs on effect-payload-nested defs), and the doc claim has been narrowed with an explicit `KNOWN GAP` naming every carrier. Worth tracking rather than closing, because the trigger is a *parser* change that compiles silently and fails no existing test.

## Summary by CodeRabbit

- **Bug Fixes**
  - Correctly excludes activated abilities that move cards to or from a library from mana-ability classification.
  - Distinguishes library movement from revealing, reordering, and other non-moving effects.
  - Preserves renewable mana-source classification where appropriate.
  - Improves handling of nested costs, effects, triggers, replacements, and delayed resolutions.
- **Tests**
  - Added comprehensive coverage for library-moving abilities, mana-source filtering, stack behavior, and activation triggers.

### CODE REVIEW

File: crates/engine/src/types/ability_visit.rs [hunk 1]
```diff
@@ -27,34 +27,249 @@
...
+/// effect-payload-nested `AbilityDefinition`. Closing it means driving both
+/// walks from one shared carrier→nested-def list; that widens cost coverage for
+/// every existing `IncludeRegisteredLater` caller too, so it needs its own
+/// census rather than riding along here.
+///
+/// Only TOP-LEVEL cost nodes are yielded; composition (`Composite`, `OneOf`,
+/// `PerCounter`) is the consuming predicate's own recursion to do. A second
+/// wildcard-free `AbilityCost` match in this module would be a drift hazard
+/// against `visit_cost_scoped`, which is exactly the pattern this module's docs
+/// reject.
```
[the remark this answers is not in the document]
Comment (matthewevans) [reply crates/engine/src/types/ability_visit.rs:232]: Confirmed, and this matches an independent implementation review that reached the same conclusion — so I've treated it as real rather than speculative. Disposition: **documented and chartered now, fixed in an immediate follow-up**, not in this PR.

Reasoning:

- **Zero reachable instances.** `data/card-data.json` carries no costs of any type on an effect-payload-nested `AbilityDefinition`, so no card is misclassified today. Your own script's `nested-carrier cost hits = 0` agrees with my census.
- **The doc claim that was actually wrong is already fixed here.** `visit_ability_def_costs_scoped` previously claimed it yields every `AbilityCost` "in the same own-resolution tree the effect walk covers" — false. It is now narrowed to the chain-link axis, with an explicit `KNOWN GAP` block naming every carrier (`Vote` / `VoteSubject.outcome_template`, `SeparateIntoPiles`, `RevealFromHand.on_decline`, `FlipCoin`, `FlipCoins`, `FlipCoinUntilLose`, `RollDie.results`, `ChooseOneOf.branches`) plus the `Effect::PayCost` one-node delegation.
- **The real fix needs its own census.** Driving both walks from one shared carrier→nested-def list widens *cost* coverage for every existing `IncludeRegisteredLater` caller — 8 public entry points, ~52 call sites. That is a behavior change for callers unrelated to CR 605.1a and should not ride along on a classifier change.

Your second option is the right near-term mitigation and I'm taking it: a corpus assertion pinning zero costs on effect-payload-nested defs. That matters specifically because **the trigger for this gap lives in a different subsystem** — it activates when a *parser* production starts emitting a cost on a branch-carried def, which compiles silently and fails no existing test. A doc comment only reaches someone already editing the walker; a corpus test fails at the moment the shape is introduced. Precedent: `tests/integration/battlefield_entry_authority_census.rs`.

Full charter (both closure options, with the call-site census scoped) is recorded alongside the run artifacts.

[the remark this answers is not in the document]
Comment (matthewevans) [reply crates/engine/src/types/ability_visit.rs:232]: Follow-up is up: #7114 adds exactly the corpus assertion we agreed on, so no issue is needed.

`crates/engine/tests/integration/cost_walk_carrier_guard.rs` walks the export and fails if any `AbilityDefinition`-shaped node carrying a non-null `cost` sits underneath one of the eight inline branch carriers (`Vote`, `SeparateIntoPiles`, `RevealFromHand`, `FlipCoin`, `FlipCoins`, `FlipCoinUntilLose`, `RollDie`, `ChooseOneOf`). The `under_carrier` marker is sticky, so a cost anywhere in the subtree counts — the whole subtree is unreachable from the cost walk, not just the carrier's immediate child.

Two things worth flagging about how it was validated, since a zero-hit assertion is the easy thing to get wrong:

- It loads through `support::shared_card_export_json()`, which returns `None` when the gitignored export is absent. The first run passed in 0.01s — that was a self-skip, not a pass. With the real 99MB export symlinked in it runs 5.06s over 35,657 cards.
- I falsified it before trusting the green: seeding the walk as though the root were already under a carrier produced 13,511 hits with card names and JSON paths, and the failure message rendered correctly. Reverting returned it to green.

It also carries two reach guards inside the test — a minimum entry count, and an assertion that at least one node in the corpus actually carries one of the eight tags, so a future rename of the serialized tags fails loudly instead of passing vacuously.

The real fix (driving `visit_ability_def_costs_scoped` and `visit_effect_scoped` from one shared carrier→nested-def list) still needs the census across the 8 `IncludeRegisteredLater` entry points, and stays chartered separately.

[the remark this answers is not in the document]
Comment (matthewevans) [reply crates/engine/src/types/ability_visit.rs:232]: Correction to my previous reply, since it would have overstated the guard.

I described #7114 as a Rust integration test that "will fail when the parser first emits a cost on an effect-payload-nested `AbilityDefinition`." That was only true locally. The full export is gitignored and absent from CI's `Rust tests` job, so a test reading it through `support::shared_card_export_json()` self-skips there — the header of `scripts/check-test-card-data-load.sh` says so outright: such tests are "invisible in CI yet bloat every local and Tilt `test-engine` run."

That matters because of what the guard is for. The trigger is a parser production emitting a cost on a branch-carried definition, and that author's change arrives through CI. A CI-invisible guard would have been green for exactly the person it exists to stop, while costing ~5s on every local test run.

So #7114 is now a `card-data-gate` step running `scripts/cost_walk_carrier_census.py --check`, placed immediately after `draw_replacement_census.py --corpus` — same reasoning, same job, unconditional (the gates-cache does not hash `scripts/**`, so a cache hit must not skip it). Missing card-data is exit 2, never a pass; that precedent script states the rule this one follows: "a gate that quietly passes when its input is missing is not a gate."

The Rust test is removed rather than kept alongside, so the predicate has a single implementation.

Falsified against the live 35,657-card export: clean run reaches 437 carrier nodes with 0 hits; seeding the walk as if the root were under a carrier yields 13,511 hits with card names and paths — the same count the Rust prototype gave under the same seed, which is what confirms the port is equivalent. Renaming the eight carrier tags and truncating the export each trip a separate reach guard.
